streams/src/test/java/org/apache/kafka/streams/integration/KStreamKTableJoinIntegrationTest.java

```diff
@@ -295,10 +314,26 @@ public Long apply(final Long value1, final Long value2) {
         consumerConfig.put(ConsumerConfig.KEY_DESERIALIZER_CLASS_CONFIG, StringDeserializer.class);
         consumerConfig.put(ConsumerConfig.VALUE_DESERIALIZER_CLASS_CONFIG, LongDeserializer.class);
 
-        final List<KeyValue<String, Long>> actualClicksPerRegion = IntegrationTestUtils.waitUntilMinKeyValueRecordsReceived(consumerConfig,
-            outputTopic, expectedClicksPerRegion.size());
-
-        assertThat(actualClicksPerRegion, equalTo(expectedClicksPerRegion));
+        TestUtils.waitForCondition(new TestCondition() {
+            @Override
+            public boolean conditionMet() {
+                return remainingExpectedResult.isEmpty();
+            }
+        }, "Never received expected result.");
+
+        final int expectedResultSize = expectedClicksPerRegion.size();
+        final int expectedNumberOfOutputs = (cacheSizeBytes == 0)
+            ? expectedResultSize
+            : Math.max(expectedResultSize, outputCounter.get());
+        final List<KeyValue<String, Long>> actualClicksPerRegion =
+            IntegrationTestUtils.waitUntilMinKeyValueRecordsReceived(
```

Comment:
Checking that all going through records are indeed produced and hence can be consumed is more strict but maybe an overkill? I'm wondering what if we just use `TestUtils.waitForCondition` to check that we have at least seen the three records in `expectedClicksPerRegion`, and throw exception if we have seen more records with the same key after we have seen these records for those keys?

Reply:
I think, this would only work if we do a loop and read a single record each time. However, to make this reliable, for each key, we would need to check all possible intermediate values (as we don't know if/when a flush might have hit). This would make the code rather complex IMHO. Just checking if the final result is correct (ie, the last value per key) seems to be sufficient and simpler.

Thoughts?

Reply:
That makes sense, I think keeping it as-is is better.